### DISCUSSION TITLE
How to turn into JSON string?

### DISCUSSION THREAD

**Discussion by dbircsak** [karljj1/kdis#31] [General]
Hi, I'm trying to figure out how to use this library. I'd like to turn an Entity_State_PDU into a "JSON string" that I can write out and read from Python, however I don't know how to access several parameters, namely variableParameters. This is the code I have:

`std::string kdisHelper::to_json(const KDIS::PDU::Entity_State_PDU &pdu) {
    std::stringstream ss;
    ss << "{\"protocolVersion\": " << pdu.GetProtocolVersion();
    ss << ", \"exerciseID\": " << (int) pdu.GetExerciseID();
    ss << ", \"pduType\": " << pdu.GetPDUType();
    ss << ", \"protocolFamily\": " << pdu.GetProtocolFamily();
    ss << ", \"timestamp\": " << pdu.GetTimeStamp().GetTime();
    ss << ", \"length\": " << pdu.GetPDULength();
    ss << ", \"pduStatus\": " << 0;
    ss << ", \"padding\": " << 0;
    ss << ", \"entityID\": {\"siteID\": " << pdu.GetEntityIdentifier().GetSiteID();
    ss << ", \"applicationID\": " << pdu.GetEntityIdentifier().GetApplicationID();
    ss << ", \"entityID\": " << pdu.GetEntityIdentifier().GetEntityID();
    ss << "}, \"forceId\": " << pdu.GetForceID();
    ss << ", \"numberOfVariableParameters\": " << 0;

    ss << ", \"entityType\": {\"entityKind\": " << pdu.GetEntityType().GetEntityKind();
    ss << ", \"domain\": " << pdu.GetEntityType().GetDomain();
    ss << ", \"country\": " << pdu.GetEntityType().GetCountry();
    ss << ", \"category\": " << (int) pdu.GetEntityType().GetCategory();
    ss << ", \"subcategory\": " << (int) pdu.GetEntityType().GetSubCategory();
    ss << ", \"specific\": " << (int) pdu.GetEntityType().GetSpecific();
    ss << ", \"extra\": " << (int) pdu.GetEntityType().GetExtra();

    ss << "}, \"alternativeEntityType\": {\"entityKind\": " << pdu.GetAltEntityType().GetEntityKind();
    ss << ", \"domain\": " << pdu.GetAltEntityType().GetDomain();
    ss << ", \"country\": " << pdu.GetAltEntityType().GetCountry();
    ss << ", \"category\": " << (int) pdu.GetAltEntityType().GetCategory();
    ss << ", \"subcategory\": " << (int) pdu.GetAltEntityType().GetSubCategory();
    ss << ", \"specific\": " << (int) pdu.GetAltEntityType().GetSpecific();
    ss << ", \"extra\": " << (int) pdu.GetAltEntityType().GetExtra();

    ss << "}, \"entityLinearVelocity\": {\"x\": " << pdu.GetEntityLinearVelocity().GetX();
    ss << ", \"y\": " << pdu.GetEntityLinearVelocity().GetY();
    ss << ", \"z\": " << pdu.GetEntityLinearVelocity().GetZ();
    ss << "}, \"entityLocation\": {\"x\": " << pdu.GetEntityLocation().GetX();
    ss << ", \"y\": " << pdu.GetEntityLocation().GetY();
    ss << ", \"z\": " << pdu.GetEntityLocation().GetZ();
    ss << "}, \"entityOrientation\": {\"psi\": " << pdu.GetEntityOrientation().GetPsiInRadians();
    ss << ", \"theta\": " << pdu.GetEntityOrientation().GetThetaInRadians();
    ss << ", \"phi\": " << pdu.GetEntityOrientation().GetPhiInRadians();
    ss << "}, \"entityAppearance\": " << pdu.GetEntityAppearance().GetData();

    ss << ", \"deadReckoningParameters\": {\"deadReckoningAlgorithm\": "
       << pdu.GetDeadReckoningParameter().GetDeadReckoningAlgorithm();
    ss << ", \"parameters\": [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]";
    ss << ", \"entityLinearAcceleration\": {\"x\": " << pdu.GetDeadReckoningParameter().GetLinearAcceleration().GetX();
    ss << ", \"y\": " << pdu.GetDeadReckoningParameter().GetLinearAcceleration().GetY();
    ss << ", \"z\": " << pdu.GetDeadReckoningParameter().GetLinearAcceleration().GetZ();
    ss << "}, \"entityAngularVelocity\": {\"x\": " << pdu.GetDeadReckoningParameter().GetAngularVelocity().GetX();
    ss << ", \"y\": " << pdu.GetDeadReckoningParameter().GetAngularVelocity().GetY();
    ss << ", \"z\": " << pdu.GetDeadReckoningParameter().GetAngularVelocity().GetZ();

    ss << "}}, \"marking\": {\"characterSet\": " << pdu.GetEntityMarking().GetEntityMarkingCharacterSet();
    ss << ", \"characters\": [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]";
    ss << "}, \"capabilities\": " << pdu.GetEntityCapabilities().GetAllFields();
    ss << ", \"variableParameters\": []}" << std::endl;
    return ss.str();
}`

I'm TRYING to turn this into something like: 

{'protocolVersion': 6, 'exerciseID': 1, 'pduType': 1, 'protocolFamily': 1, 'timestamp': 1651226583, 'length': 176, 'pduStatus': 0, 'padding': 0, 'entityID': {'siteID': 25, 'applicationID': 2, 'entityID': 1}, 'forceId': 1, 'numberOfVariableParameters': 2, 'entityType': {'entityKind': 1, 'domain': 2, 'country': 225, 'category': 1, 'subcategory': 9, 'specific': 10, 'extra': 0}, 'alternativeEntityType': {'entityKind': 0, 'domain': 0, 'country': 0, 'category': 0, 'subcategory': 0, 'specific': 0, 'extra': 0}, 'entityLinearVelocity': {'x': 28.451427459716797, 'y': 53.50935363769531, 'z': 80.42235565185547}, 'entityLocation': {'x': -2397723.0428947527, 'y': -4509461.183342898, 'z': 3822916.2741852123}, 'entityOrientation': {'psi': 1.0821040868759155, 'theta': -0.9250205755233765, 'phi': 3.1415927410125732}, 'entityAppearance': 0, 'deadReckoningParameters': {'deadReckoningAlgorithm': 4, 'parameters': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], 'entityLinearAcceleration': {'x': 0.0, 'y': 0.0, 'z': 0.0}, 'entityAngularVelocity': {'x': 0.0, 'y': 0.0, 'z': 0.0}}, 'marking': {'characterSet': 0, 'characters': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]}, 'capabilities': 0, 'variableParameters': [{'recordType': 0, 'variableParameterFields1': 7.291122019561264e-304, 'variableParameterFields2': 2354979328, 'variableParameterFields3': 0, 'variableParameterFields4': 0}, {'recordType': 0, 'variableParameterFields1': 7.291122019562091e-304, 'variableParameterFields2': 251278592, 'variableParameterFields3': 0, 'variableParameterFields4': 0}]}

But no dice. Any idea how to access variableParameters in an Entity_State_PDU?

**Comment by carlocorradini** [comment]
Hello, @dbircsak
To obtain a vector of parameters, use the [`GetVariableParameters`](https://github.com/karljj1/kdis/blob/a8d0c4e860b42f3707e39b044f0e57c631bd313b/include/KDIS/PDU/Entity_Info_Interaction/Entity_State_PDU.hpp#L317) method of `Entity_State_PDU`

PS: I recommend using [`nlohmann::json`](https://json.nlohmann.me) and its incredibly helpful [`Arbitrary Type Conversions`](https://json.nlohmann.me/features/arbitrary_types/#how-do-i-convert-third-party-types) feature if you wish to use _JSON_ in _CPP_

**Comment by dbircsak** [comment]
Thanks. Do you know if there's any examples of how to make and attach a Variable Parameter such as an AttachedPart? The Wiki doesn't cover: https://sourceforge.net/p/kdis/wiki/Sending_a_Entity_State_PDU_Tutorial_Advanced/#articulation-parameters

**Reply by carlocorradini** [reply]
@karljj1 surely can help 🥳

**Reply by karljj1** [reply]
Its been over 10 years since I did this so I'm a bit rusty. 
There's 3 types you can use:
- ArticulatedParts
- AttachedPart
- SeparationPart

The file AttachedPart.hpp explains what each field is.
For an attached part it would be something like:

```c++
    AttachedPart* part = new AttachedPart();
    part->SetAttachedPartParameterType(AttachedPartParameterType::LeftWingStations);
    part->SetAttachedPartType(...);// The type. E.G missile
    Entity.AddVariableParameter(part);
```

**Reply by dbircsak** [reply]
Thanks. We ended up going w/ ArticulatedPart and it worked well. We were sending it from Python and we need to re-create the long long it was stored as in Python (basically print what the Python dict looked like in C++). This is what we ended up with:

`ss << ", \"variableParameters\": [";
bool print_comma = false;
const std::vector<KDIS::DATA_TYPE::VarPrmPtr> &varPrmVec = pdu.GetVariableParameters();
for (KDIS::DATA_TYPE::VarPrmPtr varPrm : varPrmVec) {
	if (varPrm->GetVariableParameterType() != KDIS::DATA_TYPE::ENUMS::VariableParameterType::ArticulatedPartType)
		continue;

	if (print_comma)
		ss << ", ";

	const KDIS::DATA_TYPE::ArticulatedPart *vp0 = dynamic_cast<const KDIS::DATA_TYPE::ArticulatedPart *>(varPrm.GetPtr());
	ss << "{\"recordType\": " << KDIS::DATA_TYPE::ENUMS::VariableParameterType::ArticulatedPartType;
	ss << ", \"changeIndicator\": " << (int) vp0->GetParameterChangeIndicator();
	ss << ", \"partAttachedTo\": " << vp0->GetAttachementID();
	ss << ", \"parameterType\": " << vp0->GetTypeVariant();
	union _twoFloatToLL {
		float f[2]; // Size 4 x 2
		long long ll; // Size 8
	} twoFloatToLL;
	twoFloatToLL.f[0] = 0;
	twoFloatToLL.f[1] = vp0->GetValue();
	ss << ", \"parameterValue\": " << twoFloatToLL.ll;
	ss << "}";

	print_comma = true;
}
ss << "]}" << std::endl;
`

**Reply by carlocorradini** [reply]
@dbircsak Can we close this? 🤗
